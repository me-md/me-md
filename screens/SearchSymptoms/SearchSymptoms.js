import React, { useState, useEffect } from 'react';
import { Dimensions, StyleSheet, Text, View } from 'react-native';
import { Body, Card, CardItem, Input, Item } from 'native-base';
import { Entypo, AntDesign, Ionicons } from '@expo/vector-icons';
import { LinearGradient } from 'expo-linear-gradient';
import { ScrollView } from 'react-native-gesture-handler';
import { searchAllSymptoms } from '../../utils/apiCalls/SymptomsAndRiskFactors/searchAllSymptoms';
import { cleanInitialUserReport } from '../../utils/helpers/helpers';

const height = Dimensions.get('window').height;

export default function SymptomsScreen({ navigation }) {
  const {
    age,
    location,
    presentFactors,
    sex,
    stateAbbreviation
  } = navigation.state.params;
  const [symptoms, setSymptoms] = useState([]);
  const [searchResults, setSearchResults] = useState([]);

  const searchSymptoms = async text => {
    if (text.length > 2) {
      const searchResults = await searchAllSymptoms(text);
      setSearchResults(searchResults.data);
    } else {
      setSearchResults([]);
    }
  };

  const displaySelectedSymptoms =
    symptoms.map((symptom, index) => {
      return (
        <CardItem key={index} style={styles.selectedSymptom}>
          <Text style={{ fontSize: 14 }}>{symptom.common_name}</Text>
          <AntDesign
            name='closecircle'
            style={styles.delete}
            id={symptom.id}
            size={26}
            color='red'
            onPress={() => {
              findSymptom(symptom)
            }}
          />
        </CardItem>)
    })

  const findSymptom = result => {
    let existingSymptom = symptoms.find(symptom => symptom.id == result.id);
    existingSymptom ? removeSymptom(result) : addSymptom(result);
  };

  const addSymptom = result => {
    let found = searchResults.find(symptom => symptom.id == result.id);
    setSymptoms([...symptoms, { ...result, present: true }]);
  };

  const removeSymptom = result => {
    let filteredSymptoms = symptoms.filter(symptom => symptom.id !== result.id);
    setSymptoms(filteredSymptoms);
  };

  const sendInitialSymptoms = () => {
    const userInfo = { age, presentFactors, symptoms, sex };
    let cleanedData = cleanInitialUserReport(userInfo);
    navigation.push('SymptomsQA', {
      cleanedData,
      location,
      stateAbbreviation
    });
  };

  const displaySymptoms = searchResults.map((result, index) => {
    let found = symptoms.find(symptom => symptom.id == result.id);
    return (
      <View key={index} style={styles.checkboxes}>
        <Text style={styles.symptomText}>{result.common_name}?</Text>
        <AntDesign
          name={found ? 'checkcircle' : 'pluscircleo'}
          id={result.id}
          style={styles.add}
          size={26}
          color={found ? 'green' : 'black'}
          onPress={() => {
            findSymptom(result);
          }}
        />
      </View>
    );
  });

  return (
    <View style={styles.container}>
      <LinearGradient
        style={styles.gradientBackground}
        colors={['#004EFF', '#88CCF1']}
      >
        <View style={styles.contentContainer}>
          <Entypo
            name='chevron-thin-up'
            size={36}
            color='white'
            onPress={() => navigation.goBack()}
          />
          <Text style={styles.title}>Symptoms</Text>
          <Item style={styles.searchBox}>
            <Input
              placeholder='Search all symptoms'
              placeholderTextColor='#f8f8f8'
              style={styles.input}
              onChangeText={text => searchSymptoms(text)}
            />
            <Ionicons
              active
              name='ios-search'
              size={36}
              style={styles.searchIcon}
            />
          </Item>
          <Text style={styles.hint}>Select all that apply (at least 3):</Text>
          <ScrollView style={styles.scroll}>
            <View style={styles.searchResultsContainer}>
              <Card style={styles.selectedContainer}>
                {displaySelectedSymptoms}
              </Card>
            </View>
            <View style={styles.searchResults}>{displaySymptoms}</View>
          </ScrollView>
          {symptoms.length >= 3 && (
            <Entypo
              name='chevron-thin-down'
              size={36}
              color='white'
              onPress={() => sendInitialSymptoms()}
            />
          )}
        </View>
      </LinearGradient>
    </View>
  );
}

const styles = StyleSheet.create({
  gradientBackground: {
    alignItems: 'center',
    flex: 1
  },
  searchIcon: {
    color: 'white'
  },
  searchBox: {
    marginBottom: height * 0.01,
    marginTop: height * 0.01,
    width: '75%'
  },
  container: {
    flex: 1,
    width: '100%'
  },
  contentContainer: {
    alignItems: 'center',
    flex: 1,
    justifyContent: 'center',
    marginBottom: height * 0.06,
    marginTop: height * 0.06,
    width: '100%'
  },
  scroll: {
    marginTop: height * 0.01,
    width: '100%'
  },
  input: {
    color: 'white',
    fontSize: 24,
    fontWeight: 'bold'
  },
  hint: {
    color: 'white',
    fontSize: 20
  },
  title: {
    color: 'white',
    fontSize: 36,
    fontWeight: 'bold',
    paddingTop: height * 0.025
  },
  searchResultsContainer: {
    marginLeft: '5%',
    width: '100%'
  },
  selectedContainer: {
    borderColor: 'rgba(256, 256, 256, 0)',
    backgroundColor: 'rgba(256, 256, 256, 0)',
    flexDirection: 'row',
    flexWrap: 'wrap',
    width: '80%'
  },
  selectedSymptom: {
    backgroundColor: 'rgba(256, 256, 256, 0.9)',
    borderRadius: 50,
    height: height * 0.08,
    flexDirection: 'row',
    justifyContent: "space-between",
    marginLeft: height * 0.01,
    marginTop: height * 0.01,
    padding: height * 0.01,
  },
  searchResults: {
    alignSelf: 'center',
    width: 0,
    flexGrow: 1,
    flex: 0.5,
    marginTop: height * 0.02,
    padding: 0,
    width: '90%'
  },
  symptomText: {
    fontSize: 18,
    width: '70%'
  },
  checkboxes: {
    alignItems: 'center',
    backgroundColor: 'rgba(256, 256, 256, 0.7)',
    borderColor: 'grey',
    borderRadius: 15,
    height: height * 0.1,
    flexDirection: 'row',
    justifyContent: "space-between",
    marginTop: height * 0.02,
    padding: 10
  },
  delete: {
    marginLeft: 10
  }
});
