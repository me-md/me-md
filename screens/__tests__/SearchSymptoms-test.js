import React from 'react';
import NavigationTestUtils from 'react-navigation/NavigationTestUtils';
import renderer from 'react-test-renderer';

import SearchSymptoms from '../../screens/SearchSymptoms/SearchSymptoms';

jest.mock('expo', () => ({
  AppLoading: 'AppLoading'
}));

describe('SearchSymptoms', () => {
  jest.useFakeTimers();

  beforeEach(() => {
    NavigationTestUtils.resetInternalState();
  });

  it(`should match a snapshot`, () => {
    const mockNavigation = {
      state: {
        params: {
          sex: 'male',
          age: 24,
          stateAbbreviation: 'CO',
          location: {
            lat: 121,
            long: 25
          },
          presentFactors: [{}, {}, {}]
        }
      }
    };

    const tree = renderer
      .create(<SearchSymptoms navigation={mockNavigation} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
