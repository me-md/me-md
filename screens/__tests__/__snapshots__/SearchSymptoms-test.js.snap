// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchSymptoms should match a snapshot 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "width": "100%",
    }
  }
>
  <ViewManagerAdapter_ExpoLinearGradient
    proxiedProperties={
      Object {
        "colors": Array [
          4278210303,
          4287155441,
        ],
        "endPoint": undefined,
        "locations": undefined,
        "startPoint": undefined,
      }
    }
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
          "justifyContent": "center",
          "marginBottom": 80.03999999999999,
          "marginTop": 80.03999999999999,
          "width": "100%",
        }
      }
    >
      <Text />
      <Text
        style={
          Object {
            "color": "white",
            "fontSize": 53.36,
            "fontWeight": "bold",
            "paddingTop": 13.34,
          }
        }
      >
        Search Symptoms
      </Text>
      <View
        accessible={true}
        focusable={false}
        isTVSelectable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "transparent",
            "borderBottomWidth": 1,
            "borderColor": "#D9D5DC",
            "borderLeftWidth": 0,
            "borderRadius": undefined,
            "borderRightWidth": 0,
            "borderTopWidth": 0,
            "borderWidth": undefined,
            "flexDirection": "row",
            "marginBottom": 13.34,
            "marginLeft": 2,
            "marginTop": 13.34,
            "opacity": 1,
            "width": "75%",
          }
        }
      >
        <TextInput
          allowFontScaling={true}
          editable={true}
          onChangeText={[Function]}
          placeholder="Search all symptoms"
          placeholderTextColor="#f8f8f8"
          rejectResponderTermination={true}
          style={
            Array [
              Object {
                "color": "#000",
                "flex": 1,
                "fontSize": 17,
                "height": 50,
                "paddingLeft": 5,
                "paddingRight": 5,
                "top": 1.5,
              },
              Object {
                "color": "white",
                "fontSize": 37.5,
                "fontWeight": "bold",
              },
            ]
          }
          underlineColorAndroid="rgba(0,0,0,0)"
        />
        <Text />
      </View>
      <Text
        style={
          Object {
            "color": "white",
            "fontSize": 20,
          }
        }
      >
        Select all that apply (at least 3):
      </Text>
      <RCTScrollView
        collapsable={false}
        onGestureHandlerEvent={[Function]}
        onGestureHandlerStateChange={[Function]}
        style={
          Object {
            "marginTop": 13.34,
            "width": "100%",
          }
        }
      >
        <View>
          <View
            style={
              Object {
                "marginLeft": "5%",
                "width": "100%",
              }
            }
          >
            <View
              style={
                Array [
                  Object {
                    "backgroundColor": "#fff",
                    "borderBottomWidth": 0.5,
                    "borderColor": "#ccc",
                    "borderLeftWidth": 0.5,
                    "borderRadius": 2,
                    "borderRightWidth": 0.5,
                    "borderTopWidth": 0.5,
                    "elevation": 3,
                    "flexWrap": "nowrap",
                    "marginBottom": 5,
                    "marginLeft": 2,
                    "marginRight": 2,
                    "marginTop": 5,
                    "shadowColor": "#000",
                    "shadowOffset": Object {
                      "height": 2,
                      "width": 0,
                    },
                    "shadowOpacity": 0.1,
                    "shadowRadius": 1.5,
                  },
                  Object {
                    "backgroundColor": "rgba(256, 256, 256, 0)",
                    "borderColor": "rgba(256, 256, 256, 0)",
                    "flexDirection": "row",
                    "flexWrap": "wrap",
                    "width": "80%",
                  },
                ]
              }
            />
          </View>
          <View
            style={
              Object {
                "alignSelf": "center",
                "flex": 0.5,
                "flexGrow": 1,
                "marginTop": 26.68,
                "padding": 0,
                "width": 675,
              }
            }
          />
        </View>
      </RCTScrollView>
    </View>
  </ViewManagerAdapter_ExpoLinearGradient>
</View>
`;
